tensorflow/contrib/avro/ops/avro_record_dataset.cc

```diff
@@ -0,0 +1,419 @@
+filenames: A scalar or vector containing the name(s) of the file(s) to be
+  read.
+schema: A string used that is used for schema resolution.
+)doc");
+
+// This class represents the avro reader options
+class AvroReaderOptions {
+ public:
+  // Creates avro reader options with the given schema and buffer size.
+  //
+  static AvroReaderOptions CreateAvroReaderOptions(const string& schema,
+                                                   int64 buffer_size) {
+    AvroReaderOptions options;
+    options.schema = schema;
+    options.buffer_size = buffer_size;
+    return options;
+  }
+  string schema;
+  int64 buffer_size =
+      256 * 1024;  // 256 kB as default but this can be overwritten by the user
+};
+
+void AvroFileReaderDestructor(avro_file_reader_t reader) {
+  // I don't think we need the CHECK_NOTNULL
+  CHECK_GE(avro_file_reader_close(reader), 0);
+}
+
+void AvroSchemaDestructor(avro_schema_t schema) {
+  // Confusingly, it appears that the avro_file_reader_t creates its
+  // own reference to this schema, so the schema is not really
+  // "uniquely" owned...
+  CHECK_GE(avro_schema_decref(schema), 0);
+};
+
+void AvroValueInterfaceDestructor(avro_value_iface_t * iface)  {
+  avro_value_iface_decref(iface);
+}
+
+
+// This reader is not thread safe
+class SequentialAvroRecordReader {
+ public:
+  // Construct a sequential avro record reader
+  //
+  // 'file' is the random access file
+  //
+  // 'file_size' is the size of the file
+  //
+  // 'filename' is the name of the file
+  //
+  // 'options' are avro reader options
+  //
+  SequentialAvroRecordReader(RandomAccessFile* file, const uint64 file_size,
+                             const string& filename,
+                             const AvroReaderOptions& options =
+                                 AvroReaderOptions())
+    : initialized_(false),
+      filename_(filename),
+      file_buffer_(file_size, '\0'),
+      input_buffer_size_(options.buffer_size),
+      input_buffer_(new io::InputBuffer(file, options.buffer_size)),
+      reader_schema_str_(options.schema),
+      file_reader_(nullptr, AvroFileReaderDestructor),
+      reader_schema_(nullptr, AvroSchemaDestructor),
+      writer_schema_(nullptr, AvroSchemaDestructor),
+      p_reader_iface_(nullptr, AvroValueInterfaceDestructor),
+      p_writer_iface_(nullptr, AvroValueInterfaceDestructor) { }
+  virtual ~SequentialAvroRecordReader() {
+    // Guard against clean-up of non-initialized instances
+    if (initialized_) {
+      avro_value_decref(&reader_value_);
+      avro_value_decref(&writer_value_);
+    }
+  }
+  // Reads the next record into the string record
+  //
+  // 'record' pointer to the string where to load the record in
+  //
+  // returns Status about this operation
+  //
+  Status ReadRecord(string* record) {
+    bool at_end =
+      avro_file_reader_read_value(file_reader_.get(), &writer_value_) != 0;
+    // Are writer_value_ and reader_value_ aliases to the same thing???
```

Comment:
I'm confused about this comment, is the issue that you would do something else if you could tell they're the same thing?

Reply:
I've removed the comment.  
For clarification: They are when there is no schema resolution (otherwise not).